src/wayland-eglsurface.c

```diff
@@ -286,8 +290,21 @@ wlEglSendDamageEvent(WlEglSurface *surface, struct wl_event_queue *queue)
                           surface->dy);
     }
 
-    wl_surface_damage(surface->wlSurface, 0, 0,
-                      surface->width, surface->height);
+    if (n_rects > 0 &&
+        (wl_proxy_get_version((struct wl_proxy *)surface->wlSurface) >=
+         WL_SURFACE_DAMAGE_BUFFER_SINCE_VERSION)) {
+        for (i = 0; i < n_rects; i++) {
+            const EGLint *rect = &rects[i * 4];
+            // Coordinate systems are flipped between eglSwapBuffersWithDamage
+            // and wl_surface_damage_buffer, so invert Y values.
+            int inv_y = surface->height - (rect[1] + rect[3]);
+            wl_surface_damage_buffer(surface->wlSurface, rect[0], inv_y, rect[2], rect[3]);
+        }
+    } else {
+        wl_surface_damage(surface->wlSurface, 0, 0, UINT32_MAX, UINT32_MAX);
```

Comment:
The width and height parameters of wl_surface.damage are of type int so this should bi INT32_MAX

Reply:
Ah, good point. I'll get that fixed.